### ISSUE TITLE
DB Browser Crashes When Importing Its Own .sql dump

### ISSUE THREAD

**Issue by gmseed** [sqlitebrowser/sqlitebrowser#1045]
Hi

I can get DB Browser to crash with the following steps:

1) Open a 120MB .db database.
2) Export the db to .sql using the File\Export option.
3) Close the project and create a new database by clicking the New Database button.
4) Open the Execute SQL tab
5) Click the icon for loading a .sql file
6) Click the > arrow to execute the .sql
7) Crash

My question is: what is the upperbound MB size of a .sql file that can be imported and has 120MB exceeded it?

Thanks
Graham

**Comment by justinclift** [comment]
Oops, just saw your mailing list message too. :wink:

Which version of our software are you using, and on what OS?

People have reported using multi-GB files with DB4S (our acronym for DB Browser for SQLite), but I'm not remembering what kind of sizes people have reported for the SQL export/import part of things.

**Comment by gmseed** [comment]
Version 3.9.1

Qt Version 5.7.0

SQLCipher Version 3.11.0

Win7-32bit

**Comment by justinclift** [comment]
Interesting.  Would you be ok to try this ~recent nightly build instead?

  &nbsp; &nbsp; https://nightlies.sqlitebrowser.org/win32/DB%20Browser%20for%20SQLite-2017-05-19-win32.exe

Note - ignore newer ones, there's a crash bug after ~19th May or so which we haven't yet tracked down. :wink:

**Comment by chrisjlocke** [comment]
Also, have you tried File | Import | Database from SQL file...

![image](https://user-images.githubusercontent.com/7872218/27288622-2fbcbad0-54ff-11e7-87c1-2add9994e5ff.png)

**Comment by gmseed** [comment]
#justinclift:
1) Downloaded: DB Browser for SQLite-2017-05-19-win32.exe
2) Clicked No when asked if want to uninstall "old" version
3) Click Next and Agree buttons
4) Click Browse...
5) Installation hangs!!!!!

Graham

**Comment by justinclift** [comment]
Gah.  Well, this is starting out bad and getting worse! :wink:

Ok, maybe let it uninstall the old version instead.  That might be causing the hang. :smile:

**Comment by chrisjlocke** [comment]
Have you 7zip?  This can 'unpack' the files.  DB4S is happy to run 'portable' (in the sense that it doesn't need to be installed).  The 'installation hangs' could be Windows popping up a shield in the taskbar (the consent thingy), but you do sound like you know what you're doing, so I don't like saying, "are you sure you've checked the taskbar....." ;) 🐔

**Comment by gmseed** [comment]
Although I selected "No" to not uninstall the previous version it DID uninstall it as I can't see the previous version. Bizarre.

**Comment by chrisjlocke** [comment]
Thats weird, as I install the nightlies regularly, and its not done that on my machine.

If the installation still hangs, just extract this ZIP.
http://remote.mumfordwood.com:56082/db4s.zip

**Comment by chrisjlocke** [comment]
@gmseed - any chance you could email me the file causing you problems?  I presume it should ZIP up to something smaller than 120 MB?

**Comment by gmseed** [comment]
@chrisjlocke - thanks but as I said in a previous post I've given up with this app. I moved to using SQLite Expert Personal. I had too many issues using this app. Also, the .db file is confidential and can't post. I'm sure similar results would be experience wit other large files, which could be easily created by repeatedly looping over a set of values and adding. Also, be shortly moving to using HSQLDB, which has a neat and robust browser.

**Comment by justinclift** [comment]
No worries.  Very much understand how frustrating it gets when something-that-needs-to-work... just doesn't. ☹️ 

@chrisjlocke If you've the time/interest to investigate more, maybe grab some of the larger databases from dev1, export them to sql, and see if they crash when being imported back?

* https://dev1.dbhub.io/datasciesotist/DMOZ_category.db
* https://dev1.dbhub.io/justinclift/spatial56day.db

**Comment by chrisjlocke** [comment]
No problem.  Will close this issue on the request of the originator, but will reopen another one if I find out anything.

**Comment by chrisjlocke** [comment]
Not sure about the crashing, but I'm assuming the delay is due to the individual transactions.  I posted ages ago about importing taking ages... can't remember if that got fixed.  I'll take another look.

**Comment by justinclift** [comment]
@chrisjlocke Hmmm, I'm more thinking we should leave this open until we've had a chance to try out some import/export of larger sized files.  It's possible something has gone weird that causes the problem in a generic way, which would let us capture and fix it.  If it's something specific to @gmseed's data set though - which we don't have access to - we might as well close it as that's one challenge too far at the moment. :wink:

**Comment by chrisjlocke** [comment]
The importing has been overhauled considerably in the latest nightlies, so I'll close this.
@gmseed - if you're able to try one of the latest nightlies, importing should be OOOOOOODLES quicker by a magnitude of MILLLLLLIONS.*

\* Not guaranteed.  

If you're still getting an issue whatsoever, please comment back (you can still comment on closed issues)

**Comment by justinclift** [comment]
Hmmm, pretty sure the importing has only been changed for CSV imports, not SQL imports.  It's a different set of code.

**Comment by chrisjlocke** [comment]
Doh, good point.  Sorry!

**Comment by justinclift** [comment]
No worries at all. :smile:

**Comment by chrisjlocke** [comment]
Lol.  Importing a 52 MB sql file is taking a while.  After 15 minutes, I'm still on 0%.

![image](https://user-images.githubusercontent.com/7872218/52628413-83fc1100-2eaf-11e9-9900-c135f3347bd2.png)

I assume its processing each line in its own transaction.  I know there is multithreading code in the CSV import, but in the 'short term', wrapping up the import in just one transaction should speed this up immensely.

**Comment by justinclift** [comment]
Hmmm, 52MB should happen pretty quickly.  As in a few seconds. :frowning:

**Comment by justinclift** [comment]
Oh... that's right.  It's a different bit of code, so maybe not.  Oops.

**Comment by MKleusberg** [comment]
Yeah, the code is far from ideal for files that large. We could think about rewriting this piece of code. However, I don't think the performance gain would be as large as with the CSV import. We're by the way only using one transaction already, so no easy way to speed it up like this :wink: It's also interesting to see that the performance varies a lot depending on the structure of the database, i.e. many short statements are much more problematic compared to few long statements it seems.

**Comment by MKleusberg** [comment]
Looking closer at this... well, the performance probably can be improved a bit. Somebody wants to prepare a benchmark while I try to speed this function up? :smile:

**Comment by chrisjlocke** [comment]
I had a 52 MB file the other day, which I had to cancel in the end.  I'll reduce that down tomorrow and give some timings, and how many lines in the file.  I wanted to compare whether multiple rows per insert was quicker/slower as well.

**Comment by justinclift** [comment]
This might be useful:

&nbsp; &nbsp; https://nightlies.sqlitebrowser.org/example_datasets/sql_benchmarking/

There are two .sql files in there, generated by exporting the [UK postcodes database](https://dbhub.io/justinclift/National%20Statistics%20Postcode%20Lookup%20UK.sqlite).

One was exported with the multiple values per insert enabled, and one with it disabled.  The files are compressed with xz, which is readily available on *nix, and (from memory) 7zip on Windows handles ok too. :smile:

**Comment by justinclift** [comment]
Noticing some interesting messages on the terminal when trying to import the "single rows" one using a local build (few days old) from the `v3.11.x` branch.  It also crashes not too long into the import:

```
"didn't receive any output from pragma defer_foreign_keys"
...
terminate called after throwing an instance of 'std::bad_alloc'
  what():  std::bad_alloc
[1]+  Aborted                 (core dumped) sqlitebrowser  (wd: ~/Databases)
```

This build uses the (ancient) version of SQLite3 in CentOS 7:

![db4s-issue_1045-screenshot1](https://user-images.githubusercontent.com/406299/52759379-ba788e00-305f-11e9-83d7-bcc80137a32a.png)

The crash in this instance isn't from running out of ram.  This desktop has 24GB (ECC) ram, and the DB4S process is only using ~6GB.  So, plenty of free ram available.

I might try building with a newer SQLite and/or our `master` branch, just to see if it makes a difference.

**Comment by chrisjlocke** [comment]
OK, so I was using the 'DB4S download stats' database.  Exported that to a 52 MB .sql file, which contained **736,920** insert statements.  I left DB4S for approx. 15mins, and it was still showing 0%.  I'll leave it lunchtime (1hr) to see what it does.

I wrote a quick VB program which simply read the .sql file, and executed the statement.  That processed the .sql file in under two minutes.

![image](https://user-images.githubusercontent.com/7872218/52783621-60b99900-304a-11e9-8a89-c398e653084a.png)

**Comment by justinclift** [comment]
Yeah, mine test crashed in less than 1 or 2 mins.  Should probably go looking for the core dump it mentioned, and see what a backtrace shows up.

**Comment by justinclift** [comment]
Core dump located.  Installing debugging symbols for things now.  **Might** be able to get something useful out of this. :smile:

**Comment by justinclift** [comment]
Backtrace of the core dump.  May turn out to be useful after all:

```
$ gdb ~/bin/sqlitebrowser core.8519 
GNU gdb (GDB) Red Hat Enterprise Linux 7.6.1-114.el7
Copyright (C) 2013 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.  Type "show copying"
and "show warranty" for details.
This GDB was configured as "x86_64-redhat-linux-gnu".
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>...
Reading symbols from /home/jc/bin/sqlitebrowser...(no debugging symbols found)...done.
[New LWP 8519]
[New LWP 8521]
[New LWP 8527]
[New LWP 8520]
[New LWP 8523]
[New LWP 8522]
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib64/libthread_db.so.1".
Core was generated by `sqlitebrowser'.
Program terminated with signal 6, Aborted.
#0  0x00007f8c02541207 in __GI_raise (sig=sig@entry=6) at ../nptl/sysdeps/unix/sysv/linux/raise.c:55
55        return INLINE_SYSCALL (tgkill, 3, pid, selftid, sig);
(gdb) bt
#0  0x00007f8c02541207 in __GI_raise (sig=sig@entry=6) at ../nptl/sysdeps/unix/sysv/linux/raise.c:55
#1  0x00007f8c025428f8 in __GI_abort () at abort.c:90
#2  0x00007f8c02e507d5 in __gnu_cxx::__verbose_terminate_handler ()
    at ../../../../libstdc++-v3/libsupc++/vterminate.cc:95
#3  0x00007f8c02e4e746 in __cxxabiv1::__terminate (handler=<optimized out>)
    at ../../../../libstdc++-v3/libsupc++/eh_terminate.cc:38
#4  0x00007f8c02e4e773 in std::terminate () at ../../../../libstdc++-v3/libsupc++/eh_terminate.cc:48
#5  0x00007f8c02e4e993 in __cxxabiv1::__cxa_throw (obj=0x7f8bdc004020, 
    tinfo=0x7f8c030d9b00 <typeinfo for std::bad_alloc>, dest=0x7f8c02e4cce0 <std::bad_alloc::~bad_alloc()>)
    at ../../../../libstdc++-v3/libsupc++/eh_throw.cc:87
#6  0x00007f8c03187102 in qBadAlloc () at global/qglobal.cpp:3057
#7  0x00007f8c031a0fc5 in QByteArray::QByteArray (this=0x7ffe31bfcdb0, size=<optimized out>) at tools/qbytearray.cpp:1654
#8  0x00007f8c0339d2f8 in QUtf8::convertFromUnicode (uc=0x7f8a1772b028, len=828766129) at codecs/qutfcodec.cpp:238
#9  0x00007f8c0321281c in QString::toUtf8_helper (str=...) at tools/qstring.cpp:4667
#10 0x000000000058b9b6 in DBBrowserDB::executeMultiSQL(QString const&, bool, bool) ()
#11 0x00000000005524fe in MainWindow::importDatabaseFromSQL() ()
#12 0x000000000064014a in MainWindow::qt_static_metacall(QObject*, QMetaObject::Call, int, void**) ()
#13 0x00007f8c03370237 in QMetaObject::activate (sender=sender@entry=0x1e0f210, signalOffset=<optimized out>, 
    local_signal_index=local_signal_index@entry=1, argv=argv@entry=0x7ffe31bfd350) at kernel/qobject.cpp:3766
#14 0x00007f8c03370de7 in QMetaObject::activate (sender=sender@entry=0x1e0f210, 
    m=m@entry=0x7f8c044ca6c0 <QAction::staticMetaObject>, local_signal_index=local_signal_index@entry=1, 
    argv=argv@entry=0x7ffe31bfd350) at kernel/qobject.cpp:3628
#15 0x00007f8c03fe9312 in QAction::triggered (this=this@entry=0x1e0f210, _t1=false) at .moc/moc_qaction.cpp:371
#16 0x00007f8c03febced in QAction::activate (this=0x1e0f210, event=<optimized out>) at kernel/qaction.cpp:1163
#17 0x00007f8c04168eb2 in QMenuPrivate::activateCausedStack (this=this@entry=0x1eaf6d0, causedStack=..., 
    action=action@entry=0x1e0f210, action_e=action_e@entry=QAction::Trigger, self=self@entry=true)
    at widgets/qmenu.cpp:1324
#18 0x00007f8c0416fca7 in QMenuPrivate::activateAction (this=this@entry=0x1eaf6d0, action=action@entry=0x1e0f210, 
    action_e=action_e@entry=QAction::Trigger, self=self@entry=true) at widgets/qmenu.cpp:1401
#19 0x00007f8c04170beb in QMenu::mouseReleaseEvent (this=<optimized out>, e=0x7ffe31bfd9e0) at widgets/qmenu.cpp:2889
#20 0x00007f8c0402e065 in QWidget::event (this=this@entry=0x1eaf690, event=event@entry=0x7ffe31bfd9e0)
    at kernel/qwidget.cpp:8818
#21 0x00007f8c04172ff3 in QMenu::event (this=0x1eaf690, e=0x7ffe31bfd9e0) at widgets/qmenu.cpp:3011
#22 0x00007f8c03fef7ec in QApplicationPrivate::notify_helper (this=<optimized out>, receiver=0x1eaf690, e=0x7ffe31bfd9e0)
    at kernel/qapplication.cpp:3722
#23 0x00007f8c03ff7449 in QApplication::notify (this=<optimized out>, receiver=0x1eaf690, e=0x7ffe31bfd9e0)
    at kernel/qapplication.cpp:3198
#24 0x00007f8c033482e6 in QCoreApplication::notifyInternal2 (receiver=0x1eaf690, event=0x7ffe31bfd9e0)
    at kernel/qcoreapplication.cpp:1018
#25 0x00007f8c03ff62ec in QApplicationPrivate::sendMouseEvent (receiver=0x1eaf690, event=0x7ffe31bfd9e0, 
    alienWidget=0x0, nativeWidget=0x1eaf690, buttonDown=<optimized out>, lastMouseReceiver=..., spontaneous=true)
    at ../../src/corelib/kernel/qcoreapplication.h:233
#26 0x00007f8c04047004 in QWidgetWindow::handleMouseEvent (this=this@entry=0x1fe52f0, event=event@entry=0x7ffe31bfde40)
    at kernel/qwidgetwindow.cpp:516
#27 0x00007f8c04049925 in QWidgetWindow::event (this=0x1fe52f0, event=0x7ffe31bfde40) at kernel/qwidgetwindow.cpp:243
#28 0x00007f8c03fef7ec in QApplicationPrivate::notify_helper (this=<optimized out>, receiver=0x1fe52f0, e=0x7ffe31bfde40)
    at kernel/qapplication.cpp:3722
#29 0x00007f8c03ff6e38 in QApplication::notify (this=0x7ffe31bfe250, receiver=0x1fe52f0, e=0x7ffe31bfde40)
    at kernel/qapplication.cpp:3481
#30 0x00007f8c033482e6 in QCoreApplication::notifyInternal2 (receiver=0x1fe52f0, event=0x7ffe31bfde40)
    at kernel/qcoreapplication.cpp:1018
#31 0x00007f8c0369ed8d in QGuiApplicationPrivate::processMouseEvent (e=0x2111a90)
    at ../../src/corelib/kernel/qcoreapplication.h:236
#32 0x00007f8c036a0ae5 in QGuiApplicationPrivate::processWindowSystemEvent (e=<optimized out>)
    at kernel/qguiapplication.cpp:1733
#33 0x00007f8c0367f10b in QWindowSystemInterface::sendWindowSystemEvents (flags=...)
    at kernel/qwindowsysteminterface.cpp:939
#34 0x00007f8bf495f430 in userEventSourceDispatch (source=<optimized out>) at qeventdispatcher_glib.cpp:77
#35 0x00007f8bfec17049 in g_main_dispatch (context=0x7f8bec0016f0) at gmain.c:3175
#36 g_main_context_dispatch (context=context@entry=0x7f8bec0016f0) at gmain.c:3828
#37 0x00007f8bfec173a8 in g_main_context_iterate (context=context@entry=0x7f8bec0016f0, block=block@entry=1, 
    dispatch=dispatch@entry=1, self=<optimized out>) at gmain.c:3901
#38 0x00007f8bfec1745c in g_main_context_iteration (context=0x7f8bec0016f0, may_block=1) at gmain.c:3962
#39 0x00007f8c03396abc in QEventDispatcherGlib::processEvents (this=0x1aeb320, flags=...)
    at kernel/qeventdispatcher_glib.cpp:423
#40 0x00007f8c03346deb in QEventLoop::exec (this=this@entry=0x7ffe31bfe1e0, flags=..., flags@entry=...)
    at kernel/qeventloop.cpp:212
#41 0x00007f8c0334f234 in QCoreApplication::exec () at kernel/qcoreapplication.cpp:1291
#42 0x00000000004f489b in main ()
```

#6-#10 seem like the relevant pieces.

**Comment by justinclift** [comment]
Ahhh, new commit added.  I'll compile from `master` now and try again. :smile:

**Comment by MKleusberg** [comment]
I have just pushed a first commit which might improve the situation a bit. I haven't done much conclusive benchmarking myself though.

One interesting thing to keep an eye on (in the new version as well as in the old version) is the timing of the progress dialog. For large files it takes some time before it shows up. Before it shows up, the entire file is read into memory and, what takes more time, all occurrences of ```BEGIN TRANSACTION;``` and ```COMMIT;``` are removed (there are some oversight in this process too: no removing of ```END TRANSACTION``` which is a synonym for ```COMMIT;``` and no removing of e.g. ```COMMIT ;```). Only after it shows up, the processed string is actually executed.

**Comment by justinclift** [comment]
Thanks @MKleusberg, testing it now.  Still crashing:

```
"didn't receive any output from PRAGMA defer_foreign_keys"
terminate called after throwing an instance of 'std::bad_alloc'
  what():  std::bad_alloc
Aborted (core dumped)
```

Note that no progress dialog is show.  As soon as a file name is picked, DB4S appears to hang, through to when it crashes ~30 seconds (rough guess) later.

Backtrace:

```
$ gdb ~/bin/sqlitebrowser core.14348 
GNU gdb (GDB) Red Hat Enterprise Linux 7.6.1-114.el7
Copyright (C) 2013 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.  Type "show copying"
and "show warranty" for details.
This GDB was configured as "x86_64-redhat-linux-gnu".
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>...
Reading symbols from /home/jc/bin/sqlitebrowser...(no debugging symbols found)...done.
[New LWP 14348]
[New LWP 14350]
[New LWP 14349]
[New LWP 14352]
[New LWP 14354]
[New LWP 14351]
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib64/libthread_db.so.1".
Core was generated by `/home/jc/bin/sqlitebrowser'.
Program terminated with signal 6, Aborted.
#0  0x00007f3903acf207 in __GI_raise (sig=sig@entry=6) at ../nptl/sysdeps/unix/sysv/linux/raise.c:55
55        return INLINE_SYSCALL (tgkill, 3, pid, selftid, sig);
(gdb) bt
#0  0x00007f3903acf207 in __GI_raise (sig=sig@entry=6) at ../nptl/sysdeps/unix/sysv/linux/raise.c:55
#1  0x00007f3903ad08f8 in __GI_abort () at abort.c:90
#2  0x00007f39043de7d5 in __gnu_cxx::__verbose_terminate_handler ()
    at ../../../../libstdc++-v3/libsupc++/vterminate.cc:95
#3  0x00007f39043dc746 in __cxxabiv1::__terminate (handler=<optimized out>)
    at ../../../../libstdc++-v3/libsupc++/eh_terminate.cc:38
#4  0x00007f39043dc773 in std::terminate () at ../../../../libstdc++-v3/libsupc++/eh_terminate.cc:48
#5  0x00007f39043dc993 in __cxxabiv1::__cxa_throw (obj=0x7f38d8004020, 
    tinfo=0x7f3904667b00 <typeinfo for std::bad_alloc>, dest=0x7f39043dace0 <std::bad_alloc::~bad_alloc()>)
    at ../../../../libstdc++-v3/libsupc++/eh_throw.cc:87
#6  0x00007f3904715102 in qBadAlloc () at global/qglobal.cpp:3057
#7  0x00007f390472efc5 in QByteArray::QByteArray (this=0x7ffc0dd11cd0, size=<optimized out>) at tools/qbytearray.cpp:1654
#8  0x00007f390492b2f8 in QUtf8::convertFromUnicode (uc=0x7f377b24e028, len=828766129) at codecs/qutfcodec.cpp:238
#9  0x00007f39047a081c in QString::toUtf8_helper (str=...) at tools/qstring.cpp:4667
#10 0x000000000059d92c in DBBrowserDB::executeMultiSQL(QString, bool, bool) ()
#11 0x000000000055d8dd in MainWindow::importDatabaseFromSQL() ()
#12 0x0000000000655c3a in MainWindow::qt_static_metacall(QObject*, QMetaObject::Call, int, void**) ()
#13 0x00007f39048fe237 in QMetaObject::activate (sender=sender@entry=0x2d32620, signalOffset=<optimized out>, 
    local_signal_index=local_signal_index@entry=1, argv=argv@entry=0x7ffc0dd12240) at kernel/qobject.cpp:3766
#14 0x00007f39048fede7 in QMetaObject::activate (sender=sender@entry=0x2d32620, 
    m=m@entry=0x7f3905a586c0 <QAction::staticMetaObject>, local_signal_index=local_signal_index@entry=1, 
    argv=argv@entry=0x7ffc0dd12240) at kernel/qobject.cpp:3628
#15 0x00007f3905577312 in QAction::triggered (this=this@entry=0x2d32620, _t1=false) at .moc/moc_qaction.cpp:371
#16 0x00007f3905579ced in QAction::activate (this=0x2d32620, event=<optimized out>) at kernel/qaction.cpp:1163
#17 0x00007f39056f6eb2 in QMenuPrivate::activateCausedStack (this=this@entry=0x2ddb380, causedStack=..., 
    action=action@entry=0x2d32620, action_e=action_e@entry=QAction::Trigger, self=self@entry=true)
    at widgets/qmenu.cpp:1324
#18 0x00007f39056fdca7 in QMenuPrivate::activateAction (this=this@entry=0x2ddb380, action=action@entry=0x2d32620, 
    action_e=action_e@entry=QAction::Trigger, self=self@entry=true) at widgets/qmenu.cpp:1401
#19 0x00007f39056febeb in QMenu::mouseReleaseEvent (this=<optimized out>, e=0x7ffc0dd128d0) at widgets/qmenu.cpp:2889
#20 0x00007f39055bc065 in QWidget::event (this=this@entry=0x2ddb340, event=event@entry=0x7ffc0dd128d0)
    at kernel/qwidget.cpp:8818
#21 0x00007f3905700ff3 in QMenu::event (this=0x2ddb340, e=0x7ffc0dd128d0) at widgets/qmenu.cpp:3011
#22 0x00007f390557d7ec in QApplicationPrivate::notify_helper (this=<optimized out>, receiver=0x2ddb340, e=0x7ffc0dd128d0)
    at kernel/qapplication.cpp:3722
#23 0x00007f3905585449 in QApplication::notify (this=<optimized out>, receiver=0x2ddb340, e=0x7ffc0dd128d0)
    at kernel/qapplication.cpp:3198
#24 0x00007f39048d62e6 in QCoreApplication::notifyInternal2 (receiver=0x2ddb340, event=0x7ffc0dd128d0)
    at kernel/qcoreapplication.cpp:1018
#25 0x00007f39055842ec in QApplicationPrivate::sendMouseEvent (receiver=0x2ddb340, event=0x7ffc0dd128d0, 
    alienWidget=0x0, nativeWidget=0x2ddb340, buttonDown=<optimized out>, lastMouseReceiver=..., spontaneous=true)
    at ../../src/corelib/kernel/qcoreapplication.h:233
#26 0x00007f39055d5004 in QWidgetWindow::handleMouseEvent (this=this@entry=0x30e0380, event=event@entry=0x7ffc0dd12d30)
    at kernel/qwidgetwindow.cpp:516
#27 0x00007f39055d7925 in QWidgetWindow::event (this=0x30e0380, event=0x7ffc0dd12d30) at kernel/qwidgetwindow.cpp:243
#28 0x00007f390557d7ec in QApplicationPrivate::notify_helper (this=<optimized out>, receiver=0x30e0380, e=0x7ffc0dd12d30)
    at kernel/qapplication.cpp:3722
#29 0x00007f3905584e38 in QApplication::notify (this=0x7ffc0dd13140, receiver=0x30e0380, e=0x7ffc0dd12d30)
    at kernel/qapplication.cpp:3481
#30 0x00007f39048d62e6 in QCoreApplication::notifyInternal2 (receiver=0x30e0380, event=0x7ffc0dd12d30)
    at kernel/qcoreapplication.cpp:1018
#31 0x00007f3904c2cd8d in QGuiApplicationPrivate::processMouseEvent (e=0x3026290)
    at ../../src/corelib/kernel/qcoreapplication.h:236
#32 0x00007f3904c2eae5 in QGuiApplicationPrivate::processWindowSystemEvent (e=<optimized out>)
    at kernel/qguiapplication.cpp:1733
#33 0x00007f3904c0d10b in QWindowSystemInterface::sendWindowSystemEvents (flags=...)
    at kernel/qwindowsysteminterface.cpp:939
#34 0x00007f38f5eed430 in userEventSourceDispatch (source=<optimized out>) at qeventdispatcher_glib.cpp:77
#35 0x00007f39001a5049 in g_main_dispatch (context=0x7f38ec0016f0) at gmain.c:3175
#36 g_main_context_dispatch (context=context@entry=0x7f38ec0016f0) at gmain.c:3828
#37 0x00007f39001a53a8 in g_main_context_iterate (context=context@entry=0x7f38ec0016f0, block=block@entry=1, 
    dispatch=dispatch@entry=1, self=<optimized out>) at gmain.c:3901
#38 0x00007f39001a545c in g_main_context_iteration (context=0x7f38ec0016f0, may_block=1) at gmain.c:3962
#39 0x00007f3904924abc in QEventDispatcherGlib::processEvents (this=0x29f8950, flags=...)
    at kernel/qeventdispatcher_glib.cpp:423
#40 0x00007f39048d4deb in QEventLoop::exec (this=this@entry=0x7ffc0dd130d0, flags=..., flags@entry=...)
    at kernel/qeventloop.cpp:212
#41 0x00007f39048dd234 in QCoreApplication::exec () at kernel/qcoreapplication.cpp:1291
#42 0x00000000004fc43b in main ()
```

**Comment by MKleusberg** [comment]
It's probably crashing on [this line](https://github.com/sqlitebrowser/sqlitebrowser/blob/e9145a0c8e413b92d26458e89defc521059d5ea0/src/sqlitedb.cpp#L966) then. That's something I haven't changed yet, so you'll probably get the same crash with the latest commit. We'll have to see what we can do about that.

**Comment by MKleusberg** [comment]
Here's what we do: we read the entire file into a QString (which is always UTF16 encoded as far as I remember). So from whatever the file encoding is, we have a conversion to UTF16. We then convert that to an UTF8 byte array which we can pass to SQLite. That's the second conversion which crashes in your case. Probably because it is using too much memory after all (including memory fragmentation maybe).

I have an idea though which I'll try later today :smile: In the meantime it's worth checking if the import has already become faster for smaller files which don't crash the application.

**Comment by justinclift** [comment]
Cool.  I'm going to get some sleep soon, but will take a look back at this tomorrow. :smile:

**Comment by MKleusberg** [comment]
Just committed my attempt to fix this. I didn't get the crash before, but I can definitely say it's taking a lot less time now until the progress dialog shows up. So I guess that's a good sign, maybe :crossed_fingers:

**Comment by justinclift** [comment]
Cool.  Starting my day now, so will look at this shortly.  With that code piece which reads the whole SQL file into a single string first, would it be feasible to instead break it into a loop for chunking the file?

eg processing (say) 10k lines at a time

That should stop the crashing, and it might not be terribly slower either.  Obviously, that last bit would need testing. :smile:

**Comment by justinclift** [comment]
Oh... testing your new code now, with the commit that uses a QByteArray instead of a QString, no longer crashes.  Instead, the progress dialog is displayed and goes through to 100% (smoothly), importing the whole file.  At the end, a new pop up asking if we wish to save the changes happens (that could probably be skipped).

So, import no longer crashes.  The whole "single" file (database size of ~858MB) is created and looks ok.  Major win. :smile:

**Comment by justinclift** [comment]
This code change should be simple enough to add to the 3.11.x branch too yeah?  eg that'd stop the crashing (for larger files) there as well

**Comment by MKleusberg** [comment]
Wow, didn't expect it to be *that* much of a difference :smile: And yeah, should be safe enough to include it in the 3.11.1 release.

@chrisjlocke Can you try again with the latest nightly build too? Would be interesting to see how it compares to your VB program :wink:

**Comment by MKleusberg** [comment]
@justinclift 
> eg processing (say) 10k lines at a time

I though about that too but I think this would be pretty complicated and definitely not worth the effort for now. The reason is that we can't expect one SQL statement per line and we cannot simply check for a semicolon at the end of a line either because that might be a comment, a string, inside a trigger definition, etc. So there would be a lot of parsing involved to be able to load the file in chunks.

**Comment by MKleusberg** [comment]
> At the end, a new pop up asking if we wish to save the changes happens (that could probably be skipped).

That should be solved now too :smile:

**Comment by chrisjlocke** [comment]
> @chrisjlocke Can you try again with the latest nightly build too? Would be interesting to see how it compares to your VB program 

Nah, my program is better! 😉 

Well, not really.  DB4S now loads the 52 MB import in approx. 13 seconds, so a lot quicker than my reading it all line by line!

**Comment by MKleusberg** [comment]
Oh cool, I guess we can close this issue then :smile: Any objections?

@chrisjlocke The entire reading happens before the progress dialog is shown. It also seems to happen reasonably fast, so probably no point flipping between 'reading statement' and 'processing statement'.
Before my first patch, instead of actually doing the work the import was spending most of the time on checking how much work there is left to do (which coincidentally is exactly what I do as a person too). The remaining ~13 seconds are almost entirely spent inside SQLite functions now. Because we can't check the progress of these we can only calculate the progress between two statements. Now in case of the 'multiple rows' option there are only a hand full of statements in the file (one of them having 736k lines) which makes the progress bar jump around. So in short, I don't think we have any chance to change this unless SQLite adds some functions which let us peek inside its inner workings and check the progress :smile:

**Comment by chrisjlocke** [comment]
Aah,  My thinking was 'OK, I'm going to pass this SQL statement to SQLite to process.  Ooh, its quite long - contains a lot of lines.  Thats gonna take a while.  I'll just let the user know its 736,000 lines worth...' rather than a progress bar at 0%.  I can understand *why* its that - as you say, only two statements, etc.  But before we pass SQLite those statements, we know they're pretty large.

**Comment by justinclift** [comment]
Looks good to me. :smile:

@MKleusberg Should I cherry-pick the commits across to `v3.11.x`, or will you?

**Comment by MKleusberg** [comment]
@chrisjlocke Ah, I see what you mean :smile: There's another issue then. We only know how many bytes are left in total before handing them over to SQLite. But we don't know where the next statement ends because SQLite determines that for us. So we get to know whether a statement is particularly long or not only when it's too late.

@justinclift I'll cherry-pick them in a second :smile:

**Comment by chrisjlocke** [comment]
> But we don't know where the next statement ends because SQLite determines that for us. 

Well cocksticks.  I've used up all my good ideas for the day.  I'll leave this with you to resolve.  😜

**Comment by MKleusberg** [comment]
@chrisjlocke Heh heh, I guess I'll tend to the other issue then :wink:

@justinclift Cherry-picking is done :smile:

**Comment by MKleusberg** [comment]
Thanks again to @chrisjlocke for bringing this issue to my attention :smile:

**Comment by chrisjlocke** [comment]
Always a pleasure.  Now get on and fix #1725 so the nightlies can be used again.  😆 💥

**Comment by MKleusberg** [comment]
Actually, I was already working on that issue before you mentioned it :wink:

**Comment by chrisjlocke** [comment]
🥇
